lib/systemd/user-level-services.nix

```diff
@@ -0,0 +1,105 @@
+  mkVM = moduleConfig: serviceConfig: nixpkgs.lib.nixosSystem {
+    inherit system;
+    modules = [ ({ ... }: {
+
+      imports = [ moduleConfig.module ];
+
+      services.${moduleConfig.nixOsServiceName} = {
+        enable = true;
+      } // serviceConfig;
+
+    }) ];
+  };
+
+in {
+  mkActivateScript = activateScriptName:
+      { module, nixOsServiceName, primaryService, primaryPortOptionName, auxiliaryServices}@moduleConfig:
+      { backupAction ? "", checkAction ? null, restoreAction ? ""}:
+      serviceConfig:
+    let
+      vm = mkVM moduleConfig serviceConfig;
+      allServices = [primaryService] ++ auxiliaryServices;
+    in
+    pkgs.writeShellScriptBin activateScriptName ''
+      set -euo pipefail
+      export XDG_RUNTIME_DIR="/run/user/$UID"
+      mkdir -v -p "$HOME/.config/systemd/user/default.target.wants"
+      ${if backupAction != ""
+      then
+        ''
+          # Deploy may change DB schema via migration and we should be able
+          # to restore its state to a point before an update to successfully rollback
+          # failed deployment
+          echo Stopping ${primaryService}
+          systemctl --user stop ${primaryService} || true
+          echo Backing up the existing state
+          ${backupAction}
+        ''
+      else
+        ""
+      }
+
+      ${pkgs.lib.concatStringsSep "\n" (builtins.map (service:
+        let
+          serviceUnit = vm.config.systemd.units.${service};
+        in
+          ''
+            rm -v -f -- "$HOME/.config/systemd/user/${service}"
+            ln -v -s ${serviceUnit.unit}/${service} "$HOME/.config/systemd/user/${service}"
+          ''
+          +
+          (if builtins.elem "default.target" serviceUnit.wantedBy
+          then
+            ''
+              rm -v -f -- "$HOME/.config/systemd/user/default.target.wants/${service}"
+              ln -v -s "$HOME/.config/systemd/user/${service}" "$HOME/.config/systemd/user/default.target.wants"
+            ''
+          else ""
+          )
+        ) allServices)}
+      systemctl --user daemon-reload
+
+      rollback() {
+        exit_code="$?"
+        if [[ $exit_code -ne 0 ]]; then
+          echo Activation failed, restoring the previous working state
+          ${restoreAction}
+        fi
+        exit "$exit_code"
+      }
+      trap rollback EXIT
+
+      ${pkgs.lib.concatStringsSep "\n" (builtins.map (service:
+        if builtins.elem "default.target" vm.config.systemd.units.${service}.wantedBy
+        then
+          ''
+            echo Restarting ${service}
+            systemctl --user restart ${service}
+          ''
+        else ""
+      ) (auxiliaryServices ++ [primaryService]))}
+
+      ${if checkAction != null
+      then
+        ''
+          retry_count=0
+          while [[ "$retry_count" -lt 10 ]]; do
+            echo "Checking if the server is up, round $((retry_count+1))"
+            set +e
+            ${checkAction vm.config.services.${moduleConfig.nixOsServiceName}}
```

Comment:
Hmm, could you elaborate on why `checkAction` is a function, while `restoreAction` and `backupAction` are plain text?

Reply:
in the existing usages we need in some way the port where the "server" is running, and the port is not part of the module config , but of the service config (staging vs prod).

In case of restoreAction/backupAction it was not needed